man/recfmt.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH RECFMT "1" "April 2014" "recfmt 1.7" "User Commands"
.SH NAME
recfmt \- apply a template to records
.SH SYNOPSIS
.B recfmt
[\fIOPTION\fR]... [\fITEMPLATE\fR]
.SH DESCRIPTION
Apply a template to records read from standard input.
.TP
\fB\-f\fR, \fB\-\-file\fR=\fIFILENAME\fR
read the template to apply from a file.
.TP
\fB\-\-help\fR
print a help message and exit.
.TP
\fB\-\-version\fR
show version and exit.
.SH AUTHOR
Written by Jose E. Marchesi.
.SH "REPORTING BUGS"
Report bugs to: bug\-recutils@gnu.org
.br
GNU recutils home page: <http://www.gnu.org/software/recutils/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2010, 2011, 2012, 2013, 2014 Jose E. Marchesi.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B recfmt
is maintained as a Texinfo manual.  If the
.B info
and
.B recfmt
programs are properly installed at your site, the command
.IP
.B info recutils
.PP
should give you access to the complete manual.


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 2010-2014 Jose E. Marchesi.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

 Version 1.7 (25 March 2014)
* Utilities:
  . rec2csv now supports a -d option to change the delimiter
    character in CSV files.
  . It is now possible to specify several recfiles in the
    command line containing anonymous records.
* Format:
  . Support for restricting the fields that can appear in records
    with the new %allowed special field.
  . If a tab separates a field name and a field value is now
    ignored and not considered part of the field value.
* librec:
  . The current version of the library is now 1.
* Emacs mode:
  . It now possible to jump several records while navigating
    using a prefix argument to the next/previous commands.
* Documentation:
  . The examples in the manual have been improved, and well as many
    other aspects: formatting, English, etc.  Special thanks to Karl
    Berry for pointing out the problems and suggesting solutions.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.6 (03 November 2013)
* Utilities:
  . Support for joins and foreign keys.
  . Grouping records by one or more fields.
  . Sorting recors by several fields.
  . Specifying rewrite rules in field expressions.
  . Writing field values in rows.
  . Dot-notation in selection expressions.
  . Support for the >= and <= operators in selection expressions.
  . Aggregated functions: sum, min, max, avg.
  . New bash loadable builtin `readrec'.
* Format:
  . New UUID field type.
  . User provided arbitrary constraints with %constraint.
  . Pre-defined constants MIN and MAX can now be used in ranges.
* librec:
  . New high-level API: rec_db_query, rec_db_insert, rec_db_delete and
    rec_db_set.
  . Documentation in the rec.h header file extended and improved.
* Emacs mode:
  . Texinfo manual describing the usage of rec-mode.
  . Support for a mode hook.
  . Comments are now supported in records.
  . add-change-log-entry in a recfile now does the right thing.
  . Commands for copying/killing fields and records.
  . Support for undo in navigation mode.
  . Field folding.
  . Editing methods for the several field types.
  . Statistics.
* Documentation:
  . The user manual has been restructured and expanded.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.5 (13 January 2012)
* The utilities will now ask interactively for a password if it was
  not provided with the -s command line option.  This avoids security
  problems related to shell history files.
* Support for octal and hexadecimal numbers has been added.  They can
  be used in both the records and selection expressions.
* It is now possible to select a given number of random records in
  many of the utilities using the -m command line option.
* The -n option now accepts a list of indexes, supporting ranges.
* The new -U (uniq) option for recsel removes duplicated fields in the
  output records.
* The new -q (quick) option allows to quickly search for the desired
  record without having to provide a complete selection expression.
* Auto generated fields are now considered integers by default.  This
  avoids repetitive patterns in record descriptors involving %auto and
  %type.
* Tab characters are now allowed in blank lines betwwen records.
* The API in rec.h is now better documented with comments, and
  improved.
* recfix now exits with an error status if there is a parse error in
  some input file.
* The usage of the internal data structures has been _vastly_
  improved, resulting in a much faster operation.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.4 (03 November 2011)
* Support for encryption with the new %confidential special field.
* Support for named types with the new %typedef special field.
* Support for sorting with the new %sort special field.
* Recutils support for org-babel is now included in the distribution
  as ob-rec.el.  It allows to read rec data into org documents in both
  raw and table formats.
* It is now possible to use double quotes for delimiting string
  literals in selection expressions, as well as single quotes.
  Unescapted newline characters are also supported.
* The manual has been vastly improved.
* recfix now supports several operations: check, sort, encrypt and
  decrypt.
* recins now generates parseable locale independent dates when
  generating time-stamps in auto-fields.
* Lots of bug fixes.

 Version 1.3 (11 February 2011)
* A quite bad (and embarrassing) bug introduced in 1.2 has been
  fixed.  The bug was causing recins, recdel and recset to report
  random integrity failures, due to a uninitialized variable.
* New special field %size, that allows to define constraints in the
  size of record sets.
* Several fixes in the user manual.
* Little portability issue that prevented one of the tests to compile
  in macos x has been fixed.
* Translations updates.

 Version 1.2 (4 February 2011)
* New utility rec2csv, to export rec data to csv files.
* Support for auto-incremented fields and timestamps with the new
  %auto field in record descriptors.
* Support for comments in enum type descriptions.
* recset now supports a new "set or add field" operation, implementing
  the common case of setting the value of a field in case it exists,
  or else add it to the record.
* The flavor of regular expressions accepted in both selection
  expressions and type descriptions is now documented in the user
  manual.
* Translations for German, Finnish, Swedish and Dutch.
* recdel now performs an integrity check before to complete the
  deletion.
* The utilities no longer "expand" the contents of remote descriptors
  in the local ones.

utils/recfmt.c
/* -*- mode: C -*-
 *
 *       File:         recfmt.c
 *       Date:         Wed Dec 22 18:20:20 2010
 *
 *       GNU recutils - recfmt
 *
 */

/* Copyright (C) 2010-2013 Jose E. Marchesi */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <string.h>
#include <getopt.h>
#include <stdlib.h>
#include <xalloc.h>
#include <regex.h>
#include <gettext.h>
#define _(str) gettext (str)

#include <rec.h>
#include <recutl.h>

/* Forward prototypes.  */
void recfmt_parse_args (int argc, char **argv);
bool recfmt_process_data (rec_db_t db);
void recfmt_process_db (rec_db_t db, char *template);
char *recfmt_apply_template (rec_record_t record, char *template);
char *recfmt_get_subst (rec_record_t record, char *str);

/*
 * Global variables
 */

char *recfmt_template = NULL;

/*
 * Command line options management.
 */

enum
{
  COMMON_ARGS,
  FILE_ARG
};

static const struct option GNU_longOptions[] =
  {
    COMMON_LONG_ARGS,
    {"file", required_argument, NULL, FILE_ARG},
    {NULL, 0, NULL, 0}
  };

/*
 * Functions.
 */

void
recutl_print_help (void)
{
  /* TRANSLATORS: --help output, recfmt synopsis.
     no-wrap */
  printf (_("\
Usage: recfmt [OPTION]... [TEMPLATE]\n"));

  /* TRANSLATORS: --help output, recfmt arguments.
     no-wrap */
  fputs(_("\
Apply a template to records read from standard input.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, recfmt arguments.
     no-wrap */
  fputs(_("\
  -f, --file=FILENAME                 read the template to apply from a file.\n"),
        stdout);

  recutl_print_help_common ();
  puts ("");
  recutl_print_help_footer ();
}

void
recfmt_parse_args (int argc,
                   char **argv)
{
  char c;
  int ret;

  while ((ret = getopt_long (argc,
                             argv,
                             "f:",
                             GNU_longOptions,
                             NULL)) != -1)
    {
      c = ret;
      switch (c)
        {
          COMMON_ARGS_CASES
        case FILE_ARG:
        case 'f':
          {
            /* Read the template from the specified file and store it
               in recfmt_template.  */
            recfmt_template = recutl_read_file (optarg);
            if (!recfmt_template)
              {
                recutl_fatal (_("can't open file %s for reading.\n"),
                              optarg);
              }

            break;
          }
        default:
          {
            exit (EXIT_FAILURE);
          }
        }
    }

  /* See if the template is specified in the command line.  */
  if (optind < argc)
    {
      if (recfmt_template)
        {
          recutl_fatal (_("don't specify a template in the command line and -f at the same time.\n"));
        }

      if ((argc - optind) != 1)
        {
          recutl_print_help ();
          exit (EXIT_FAILURE);
        }

      recfmt_template = xstrdup (argv[optind++]);
    }
}

char *
recfmt_get_subst (rec_record_t record,
                  char *str)
{
  char *res;
  rec_sex_t sex;

  sex = rec_sex_new (false);
  if (!rec_sex_compile (sex, str))
    {
      recutl_fatal (_("invalid expression in a template slot.\n"));
    }

  res = rec_sex_eval_str (sex, record);
  if (!res)
    {
      recutl_fatal (_("error evaluating expression in a template slot.\n"));
    }

  rec_sex_destroy (sex);

  return res;
}

char *
recfmt_apply_template (rec_record_t record,
                       char *template)
{
  rec_buf_t result_buf;
  char *result;
  char *tmp;
  size_t tmp_size;
  size_t result_size;
  char *p;
  regex_t regexp;
  regmatch_t matches;
  char *subst_str;

  /* Replace occurrences of:

     {{Name[N]}}

     where Name[N] is the name of a field with an optional subscript.
     If the subscript is not present then it is assumed to be 0.  If
     the contents between {{...}} are not a field name then replace the
     slot with the empty string.
  */

  if (regcomp (&regexp, "\\{\\{" "[^}]*" "\\}\\}", REG_EXTENDED) != 0)
    {
      recutl_fatal (_("recfmt_apply_template: error compiling regexp. Please report this.\n"));
    }

  result_buf = rec_buf_new (&result, &result_size);
  p = template;
  while (*p
         && (regexec (&regexp, p, 1, &matches, 0) == 0)
         && (matches.rm_so != -1))
    {
      /* Add the prolog, if any.  */
      if (matches.rm_so > 0)
        {
          tmp = xmalloc (matches.rm_so + 1);
          memcpy (tmp, p, matches.rm_so);
          tmp[matches.rm_so] = '\0';
          rec_buf_puts (tmp, result_buf);
          free (tmp);
        }

      /* Get the match.  */
      tmp_size = matches.rm_eo - matches.rm_so - 4;
      tmp = xmalloc (tmp_size + 1);
      memcpy (tmp, p + matches.rm_so + 2, tmp_size);
      tmp[tmp_size] = '\0';

      /* Advance p. */
      p = p + matches.rm_eo;

      /* Get the substitution text and append it to the result.  */
      subst_str = recfmt_get_subst (record, tmp);
      if (subst_str)
        {
          rec_buf_puts (subst_str, result_buf);
          free (subst_str);
        }
      free (tmp);
    }

  /* Add the epilog, if any.  */
  if (*p)
    {
      rec_buf_puts (p, result_buf);
    }
  
  rec_buf_close (result_buf);

  return result;
}

void
recfmt_process_db (rec_db_t db, char *template)
{
  size_t n_rset;
  rec_rset_t rset;
  rec_record_t record;
  char *result;
  rec_mset_iterator_t iter;

  /* Iterate on all the record sets.  */
  for (n_rset = 0; n_rset < rec_db_size (db); n_rset++)
    {
      rset = rec_db_get_rset (db, n_rset);

      /* Iterate on all the records.  */

      iter = rec_mset_iterator (rec_rset_mset (rset));
      while (rec_mset_iterator_next (&iter, MSET_RECORD, (const void**) &record, NULL))
        {
          /* Apply the template to this record.  */
          result = recfmt_apply_template (record, template);
          if (result && (*result != '\0'))
            {
              printf ("%s", result);
              free (result);
            }
        }

      rec_mset_iterator_free (&iter);
    }
}

int
main (int argc, char *argv[])
{
  rec_db_t db;

  recutl_init ("recfmt");

  recfmt_parse_args (argc, argv);

  db = recutl_read_db_from_file (NULL);
  if (db && recfmt_template)
    {
      recfmt_process_db (db, recfmt_template);
    }

  return EXIT_SUCCESS;
}

/* End of recfmt.c */
